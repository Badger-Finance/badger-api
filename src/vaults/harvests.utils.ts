import { greaterThanOrEqualTo } from '@aws/dynamodb-expressions';
import {
  formatBalance,
  gqlGenT,
  keyBy,
  Vault__factory,
  VaultHarvestData,
  YieldEvent,
  YieldType,
} from '@badger-dao/sdk';
import { BadgerTreeDistribution_OrderBy, SettHarvest_OrderBy } from '@badger-dao/sdk/lib/graphql/generated/badger';
import { BigNumber, ethers } from 'ethers';

import { getDataMapper, getVaultEntityId } from '../aws/dynamodb.utils';
import { VaultDefinitionModel } from '../aws/models/vault-definition.model';
import { VaultYieldEvent } from '../aws/models/vault-yield-event.model';
import { Chain } from '../chains/config/chain.config';
import { OrderDirection } from '../graphql/generated/balancer';
import { queryPriceAtTimestamp } from '../prices/prices.utils';
import { getFullToken } from '../tokens/tokens.utils';
import { getInfuelnceVaultYieldBalance, isInfluenceVault } from './influence.utils';
import { VaultYieldItem } from './interfaces/vault-yield-item.interface';
import { VAULT_TWAY_DURATION } from './vaults.config';
import { getStrategyInfo } from './vaults.utils';
import { calculateYield } from './yields.utils';

// this allows five chunks of 10k blocks per index, approx 2 week
export const HARVEST_SCAN_BLOCK_INCREMENT = 100_000;

function grabYieldGraphTx(graphTxId?: string): string {
  if (!graphTxId) {
    return '';
  }

  if (!graphTxId.includes('-')) {
    return graphTxId;
  }

  return graphTxId.split('-')[0];
}

/**
 * Modify subgraph response to match on chain event data allowing it to fit into our estimation functions.
 * @param vault vault requesting graph data transformation
 * @param harvests harvests data retrieved from the graph
 * @param distributions distribution data retrieved from the graph
 * @returns vault harvest data in the same form as delivered via event logs
 */
function constructGraphVaultData(
  vault: VaultDefinitionModel,
  harvests: gqlGenT.SettHarvestsQuery['settHarvests'],
  distributions: gqlGenT.BadgerTreeDistributionsQuery['badgerTreeDistributions'],
): VaultHarvestData[] {
  const harvestsByTimestamp = keyBy(harvests, (harvest) => harvest.timestamp);
  const treeDistributionsByTimestamp = keyBy(distributions, (distribution) => distribution.timestamp);
  const timestamps = Array.from(
    new Set([...harvestsByTimestamp.keys(), ...treeDistributionsByTimestamp.keys()]).values(),
  );
  return timestamps.map((t) => {
    const timestamp = Number(t);
    const currentHarvests = harvestsByTimestamp.get(timestamp) ?? [];
    const currentDistributions = treeDistributionsByTimestamp.get(timestamp) ?? [];
    return {
      timestamp,
      harvests: currentHarvests.map((h) => ({
        tx: grabYieldGraphTx(h.id),
        timestamp,
        block: Number(h.blockNumber),
        token: vault.depositToken,
        amount: h.amount,
      })),
      treeDistributions: currentDistributions.map((d) => {
        const tokenAddress = d.token.id.startsWith('0x0x') ? d.token.id.slice(2) : d.token.id;
        return {
          tx: grabYieldGraphTx(d.id),
          timestamp,
          block: Number(d.blockNumber),
          token: ethers.utils.getAddress(tokenAddress),
          amount: d.amount,
        };
      }),
    };
  });
}

/**
 *
 * @param data
 * @returns
 */
function constructYieldItems(data: VaultHarvestData[]): VaultYieldItem[] {
  const recentHarvests = data.sort((a, b) => b.timestamp - a.timestamp);
  const allHarvests = recentHarvests.flatMap((h) => h.harvests.map((h) => ({ ...h, type: YieldType.Harvest })));
  const allDistributions = recentHarvests.flatMap((h) =>
    h.treeDistributions.map((d) => ({ ...d, type: YieldType.TreeDistribution })),
  );
  return allHarvests
    .concat(allDistributions)
    .filter((e) => BigNumber.from(e.amount).gt(ethers.constants.Zero))
    .sort((a, b) => b.timestamp - a.timestamp);
}

/**
 *
 * @param chain
 * @param vault
 * @param block
 * @returns
 */
async function getVaultBalance(chain: Chain, vault: VaultDefinitionModel, block: number): Promise<number> {
  const sdk = await chain.getSdk();

  let balance = 0;
  if (isInfluenceVault(vault.address)) {
    balance = await getInfuelnceVaultYieldBalance(chain, vault, block);
  } else {
    const vaultContract = Vault__factory.connect(vault.address, sdk.provider);
    const totalSupply = await vaultContract.totalSupply({ blockTag: block });
    balance = formatBalance(totalSupply);
  }

  return balance;
}

/**
 * Evaluate yield events to derivate base apr, and apy as well as apr of any emitted tokens.
 * @param chain network the vault is deployed on
 * @param vault vault requested yield event evaluation
 * @param yieldEvents events sourced from either on chain or the graph
 * @returns yield summary providing the base information for construction of yield sources
 */
async function evaluateYieldItems(
  chain: Chain,
  vault: VaultDefinitionModel,
  yieldItems: VaultYieldItem[],
): Promise<YieldEvent[]> {
  const yieldEvents: YieldEvent[] = [];
  const tokenEmissionAprs = new Map<string, number>();

  for (const item of yieldItems) {
    const block = item.block;
    const token = await getFullToken(chain, item.token);
    const amount = formatBalance(item.amount, token.decimals);
    const { price } = await queryPriceAtTimestamp(token.address, item.timestamp * 1000);

    const tokenEarned = price * amount;

    const balance = await getVaultBalance(chain, vault, item.block);
    const { price: vaultPrice } = await queryPriceAtTimestamp(vault.address, item.timestamp * 1000);
    const vaultPrincipal = vaultPrice * balance;
    const strategyInfo = await getStrategyInfo(chain, vault, { blockTag: item.block });
    const performanceScalar = 1 / (1 - strategyInfo.performanceFee / 10_000);

    const eventApr = calculateYield(vaultPrincipal, tokenEarned, VAULT_TWAY_DURATION);
    const yieldEvent: YieldEvent = {
      block,
      timestamp: item.timestamp * 1000,
      amount,
      token: token.address,
      type: item.type,
      value: vaultPrincipal,
      balance,
      earned: tokenEarned,
      apr: eventApr,
      grossApr: eventApr * performanceScalar,
      tx: item.tx,
    };
    yieldEvents.push(yieldEvent);

    if (item.type === YieldType.TreeDistribution) {
      const entry = tokenEmissionAprs.get(token.address) ?? 0;
      tokenEmissionAprs.set(token.address, entry + eventApr);
    }
  }

  return yieldEvents;
}

/**
 *
 * @param chain
 * @param vault
 * @param cutoff
 * @returns
 */
async function loadGraphYieldData(
  chain: Chain,
  vault: VaultDefinitionModel,
  cutoff: number,
): Promise<VaultHarvestData[]> {
  const { graph } = await chain.getSdk();
  const { address } = vault;
  console.log(`[${vault.name}]: Load Graph Yield Data`);
  const [vaultHarvests, treeDistributions] = await Promise.all([
    graph.loadSettHarvests({
      first: 25,
      where: {
        sett: address.toLowerCase(),
        timestamp_gt: cutoff,
      },
      orderBy: SettHarvest_OrderBy.Timestamp,
      orderDirection: OrderDirection.Asc,
    }),
    graph.loadBadgerTreeDistributions({
      first: 25,
      where: {
        sett: address.toLowerCase(),
        timestamp_gt: cutoff,
      },
      orderBy: BadgerTreeDistribution_OrderBy.Timestamp,
      orderDirection: OrderDirection.Asc,
    }),
  ]);
  const { settHarvests } = vaultHarvests;
  const { badgerTreeDistributions } = treeDistributions;
  return constructGraphVaultData(vault, settHarvests, badgerTreeDistributions);
}

/**
 *
 * @param chain
 * @param vault
 * @param lastHarvestBlock
 * @param cutoff
 * @returns
 */
async function loadEventYieldData(
  chain: Chain,
  vault: VaultDefinitionModel,
  lastHarvestBlock: number,
  cutoff: number,
): Promise<VaultHarvestData[]> {
  const { vaults } = await chain.getSdk();
  const { address, version } = vault;
  console.log(`[${vault.name}]: Load Event Yield Data`);
  try {
    const { data } = await vaults.listHarvests({
      address,
      timestamp_gte: cutoff,
      version,
      startBlock: lastHarvestBlock,
      endBlock: lastHarvestBlock + HARVEST_SCAN_BLOCK_INCREMENT,
    });
    return data;
  } catch {
    return loadGraphYieldData(chain, vault, cutoff);
  }
}

/**
 *
 * @param chain
 * @param vault
 * @param lastHarvestBlock
 * @returns
 */
export async function loadYieldEvents(
  chain: Chain,
  vault: VaultDefinitionModel,
  lastHarvestBlock: number,
): Promise<YieldEvent[]> {
  const { provider } = await chain.getSdk();
  const block = await provider.getBlock(lastHarvestBlock);
  const cutoff = block.timestamp;

  let data: VaultHarvestData[] = [];

  try {
    data = await loadGraphYieldData(chain, vault, cutoff);
  } catch (err) {
    if (isInfluenceVault(vault.address)) {
      throw err;
    } else {
      data = await loadEventYieldData(chain, vault, lastHarvestBlock, cutoff);
    }
  }

  const yieldItems = constructYieldItems(data);
  return evaluateYieldItems(chain, vault, yieldItems);
}

/**
 *
 * @param chain
 * @param vault
 * @returns
 */
export async function queryVaultYieldEvents(chain: Chain, vault: VaultDefinitionModel): Promise<VaultYieldEvent[]> {
  const mapper = getDataMapper();
  const chainAddress = getVaultEntityId(chain, vault);
  const cutoff = Date.now() - VAULT_TWAY_DURATION;

  const yieldEvents = [];
  for await (const yieldEvent of mapper.query(
    VaultYieldEvent,
    { chainAddress, timestamp: greaterThanOrEqualTo(cutoff) },
    { indexName: 'IndexYieldDataOnAddress', scanIndexForward: false },
  )) {
    yieldEvents.push(yieldEvent);
  }
  return yieldEvents;
}

/**
 *
 * @param chain
 * @param vault
 * @returns
 */
export async function queryVaultHistoricYieldEvents(
  chain: Chain,
  vault: VaultDefinitionModel,
): Promise<VaultYieldEvent[]> {
  // TODO: construct a new paginated 'all harvests' table
  // allow for page queries, timestamp based page queries
  // all pages will be the same size, new pages are only added after x entries to a given page
  // we can follow similar to the charting pattern with this data
  const mapper = getDataMapper();
  const chainAddress = getVaultEntityId(chain, vault);

  const yieldEvents = [];
  for await (const yieldEvent of mapper.query(
    VaultYieldEvent,
    { chainAddress },
    { indexName: 'IndexYieldDataOnAddress', scanIndexForward: false },
  )) {
    yieldEvents.push(yieldEvent);
  }
  return yieldEvents;
}
