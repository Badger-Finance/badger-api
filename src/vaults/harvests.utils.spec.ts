import { Vault, Vault__factory, VaultsService } from '@badger-dao/sdk';
import { BigNumber } from 'ethers';

import { Chain } from '../chains/config/chain.config';
import { TOKENS } from '../config/tokens.config';
import * as pricesUtils from '../prices/prices.utils';
import { SourceType } from '../rewards/enums/source-type.enum';
import {
  MOCK_TOKENS,
  MOCK_VAULT_DEFINITION,
  MOCK_YIELD_EVENT,
  TEST_CURRENT_BLOCK,
  TEST_CURRENT_TIMESTAMP,
} from '../test/constants';
import { mockQuery, setupMockChain } from '../test/mocks.utils';
import { mockContract } from '../test/mocks.utils/contracts/mock.contract.base';
import * as tokensUtils from '../tokens/tokens.utils';
import * as influenceUtils from '../vaults/influence.utils';
import * as vaultsUtils from '../vaults/vaults.utils';
import { loadYieldEvents, queryVaultHistoricYieldEvents, queryVaultYieldEvents } from './harvests.utils';
import { VAULT_SOURCE } from './vaults.config';
import { createYieldSource } from './yields.utils';

describe('harvests.utils', () => {
  let chain: Chain;

  beforeEach(() => {
    chain = setupMockChain();
  });

  function setupYieldMocks() {
    // setup token responses
    jest.spyOn(tokensUtils, 'getFullToken').mockImplementation(async (_c, t) => MOCK_TOKENS[t]);
    jest.spyOn(pricesUtils, 'queryPriceAtTimestamp').mockImplementation(async (token, _t, _c) => {
      const basePrice = await chain.strategy.getPrice(token);
      return {
        ...basePrice,
        updatedAt: TEST_CURRENT_TIMESTAMP,
      };
    });

    // setup contract interactions
    const vaultMock = mockContract<Vault>(Vault__factory);
    jest.spyOn(vaultMock, 'totalSupply').mockImplementation(async (_o) => BigNumber.from('10000000000000000000000000'));

    jest
      .spyOn(vaultsUtils, 'queryYieldSources')
      .mockImplementation(async (v) => [createYieldSource(v, SourceType.PreCompound, VAULT_SOURCE, 3)]);

    jest.spyOn(influenceUtils, 'getInfuelnceVaultYieldBalance').mockImplementation(async (_c, _v) => 4_500_000);
    jest.spyOn(chain.sdk.graph, 'loadSettHarvests').mockImplementation(async () => ({
      settHarvests: [
        {
          id: '0x56c8c982ce6efa45da0ff90383d42cc026397f3bceac93fa3c08d264aad1863c-106',
          timestamp: 1660107820,
          token: {
            id: '0x4e3fbd56cd56c3e72c1403e103b45db9da5b9d2b',
            name: 'Convex Token',
            symbol: 'CVX',
            decimals: '18',
            totalSupply: '52842757666481041544194387',
          },
          amount: '0',
          blockNumber: '15312417',
          strategy: { id: '0x898111d1f4eb55025d0036568212425ee2274082' },
          sett: { id: '0xfd05d3c7fe2924020620a8be4961bbaa747e6305' },
        },
        {
          id: '0x6902d5e3d0dc784dac28506ec173727111439ba5dd8a05cfb9c9a7f86137cab5-51',
          timestamp: 1661148573,
          token: {
            id: '0x4e3fbd56cd56c3e72c1403e103b45db9da5b9d2b',
            name: 'Convex Token',
            symbol: 'CVX',
            decimals: '18',
            totalSupply: '52842757666481041544194387',
          },
          amount: '0',
          blockNumber: '15388629',
          strategy: { id: '0x898111d1f4eb55025d0036568212425ee2274082' },
          sett: { id: '0xfd05d3c7fe2924020620a8be4961bbaa747e6305' },
        },
        {
          id: '0x74f208d906f022f2ed290dfc01c9b283791ae3352f74638cbdd8b6af57323304-26',
          timestamp: 1660455256,
          token: {
            id: '0x4e3fbd56cd56c3e72c1403e103b45db9da5b9d2b',
            name: 'Convex Token',
            symbol: 'CVX',
            decimals: '18',
            totalSupply: '52842757666481041544194387',
          },
          amount: '0',
          blockNumber: '15337891',
          strategy: { id: '0x898111d1f4eb55025d0036568212425ee2274082' },
          sett: { id: '0xfd05d3c7fe2924020620a8be4961bbaa747e6305' },
        },
        {
          id: '0xba80fa9bdd08636514171ee7c54d6a5420d392111c4b08fa185e9677fea5d647-44',
          timestamp: 1660800880,
          token: {
            id: '0x4e3fbd56cd56c3e72c1403e103b45db9da5b9d2b',
            name: 'Convex Token',
            symbol: 'CVX',
            decimals: '18',
            totalSupply: '52842757666481041544194387',
          },
          amount: '0',
          blockNumber: '15363244',
          strategy: { id: '0x898111d1f4eb55025d0036568212425ee2274082' },
          sett: { id: '0xfd05d3c7fe2924020620a8be4961bbaa747e6305' },
        },
      ],
    }));
    jest.spyOn(chain.sdk.graph, 'loadBadgerTreeDistributions').mockImplementation(async () => ({
      badgerTreeDistributions: [
        {
          id: '0x49730f8f1d3aa7915940ec85de53b10e5425cddbf9195c2348f779638eed14dd-85',
          timestamp: 1661253900,
          token: {
            id: '0x0xfd05d3c7fe2924020620a8be4961bbaa747e6305',
            name: 'Badger Vested Escrow Convex Token',
            symbol: 'bveCVX',
            decimals: '18',
            totalSupply: '1146328308588079160439895',
          },
          amount: '18087137228369161871035',
          blockNumber: '15396257',
          strategy: { id: '0x0000000000000000000000000000000000000000' },
          sett: { id: '0xfd05d3c7fe2924020620a8be4961bbaa747e6305' },
        },
        {
          id: '0x49730f8f1d3aa7915940ec85de53b10e5425cddbf9195c2348f779638eed14dd-92',
          timestamp: 1661253900,
          token: {
            id: '0x3472a5a71965499acd81997a54bba8d852c6e53d',
            name: 'Badger',
            symbol: 'BADGER',
            decimals: '18',
            totalSupply: '21000000000000000000000000',
          },
          amount: '8273444229204367267242',
          blockNumber: '15396257',
          strategy: { id: '0x0000000000000000000000000000000000000000' },
          sett: { id: '0xfd05d3c7fe2924020620a8be4961bbaa747e6305' },
        },
        {
          id: '0x56c8c982ce6efa45da0ff90383d42cc026397f3bceac93fa3c08d264aad1863c-105',
          timestamp: 1660107820,
          token: {
            id: '0x0x2b5455aac8d64c14786c3a29858e43b5945819c0',
            name: 'Badger Sett Convex CRV',
            symbol: 'bcvxCRV',
            decimals: '18',
            totalSupply: '9680860937976920554007',
          },
          amount: '5603203309292930442532',
          blockNumber: '15312417',
          strategy: { id: '0x898111d1f4eb55025d0036568212425ee2274082' },
          sett: { id: '0xfd05d3c7fe2924020620a8be4961bbaa747e6305' },
        },
        {
          id: '0x6902d5e3d0dc784dac28506ec173727111439ba5dd8a05cfb9c9a7f86137cab5-50',
          timestamp: 1661148573,
          token: {
            id: '0x0x2b5455aac8d64c14786c3a29858e43b5945819c0',
            name: 'Badger Sett Convex CRV',
            symbol: 'bcvxCRV',
            decimals: '18',
            totalSupply: '9680860937976920554007',
          },
          amount: '4261975371596002775428',
          blockNumber: '15388629',
          strategy: { id: '0x898111d1f4eb55025d0036568212425ee2274082' },
          sett: { id: '0xfd05d3c7fe2924020620a8be4961bbaa747e6305' },
        },
        {
          id: '0x74f208d906f022f2ed290dfc01c9b283791ae3352f74638cbdd8b6af57323304-25',
          timestamp: 1660455256,
          token: {
            id: '0x0x2b5455aac8d64c14786c3a29858e43b5945819c0',
            name: 'Badger Sett Convex CRV',
            symbol: 'bcvxCRV',
            decimals: '18',
            totalSupply: '9680860937976920554007',
          },
          amount: '4285530143667972615612',
          blockNumber: '15337891',
          strategy: { id: '0x898111d1f4eb55025d0036568212425ee2274082' },
          sett: { id: '0xfd05d3c7fe2924020620a8be4961bbaa747e6305' },
        },
        {
          id: '0x99cd95c4fd0c68c963142db17be61a8c2458ff6b5acda3aabe405a644adaee93-178',
          timestamp: 1660132736,
          token: {
            id: '0x0xfd05d3c7fe2924020620a8be4961bbaa747e6305',
            name: 'Badger Vested Escrow Convex Token',
            symbol: 'bveCVX',
            decimals: '18',
            totalSupply: '1146328308588079160439895',
          },
          amount: '18705595517466124968652',
          blockNumber: '15314213',
          strategy: { id: '0x0000000000000000000000000000000000000000' },
          sett: { id: '0xfd05d3c7fe2924020620a8be4961bbaa747e6305' },
        },
        {
          id: '0x99cd95c4fd0c68c963142db17be61a8c2458ff6b5acda3aabe405a644adaee93-185',
          timestamp: 1660132736,
          token: {
            id: '0x3472a5a71965499acd81997a54bba8d852c6e53d',
            name: 'Badger',
            symbol: 'BADGER',
            decimals: '18',
            totalSupply: '21000000000000000000000000',
          },
          amount: '8418142660217834521465',
          blockNumber: '15314213',
          strategy: { id: '0x0000000000000000000000000000000000000000' },
          sett: { id: '0xfd05d3c7fe2924020620a8be4961bbaa747e6305' },
        },
        {
          id: '0xba80fa9bdd08636514171ee7c54d6a5420d392111c4b08fa185e9677fea5d647-43',
          timestamp: 1660800880,
          token: {
            id: '0x0x2b5455aac8d64c14786c3a29858e43b5945819c0',
            name: 'Badger Sett Convex CRV',
            symbol: 'bcvxCRV',
            decimals: '18',
            totalSupply: '9680860937976920554007',
          },
          amount: '4223102994638643945171',
          blockNumber: '15363244',
          strategy: { id: '0x898111d1f4eb55025d0036568212425ee2274082' },
          sett: { id: '0xfd05d3c7fe2924020620a8be4961bbaa747e6305' },
        },
      ],
    }));
  }

  describe('loadYieldEvents', () => {
    beforeEach(setupYieldMocks);

    describe('requesting an standard vault', () => {
      it('loads yield events from the subgraph', async () => {
        const result = await loadYieldEvents(chain, MOCK_VAULT_DEFINITION, TEST_CURRENT_BLOCK);
        expect(result).toMatchSnapshot();
      });
    });

    describe('encounters an error requesting an standard vault', () => {
      it('loads yield events from the subgraph', async () => {
        jest.spyOn(VaultsService.prototype, 'listHarvests').mockImplementation(async () => {
          throw new Error('Expected test error: listHarvests');
        });
        const result = await loadYieldEvents(chain, MOCK_VAULT_DEFINITION, TEST_CURRENT_BLOCK);
        expect(result).toMatchSnapshot();
      });
    });

    describe('requesting an influence vault', () => {
      it('loads yield events from the subgraph', async () => {
        const mockVault = JSON.parse(JSON.stringify(MOCK_VAULT_DEFINITION));
        mockVault.address = TOKENS.BVECVX;
        const result = await loadYieldEvents(chain, mockVault, TEST_CURRENT_BLOCK);
        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('queryVaultYieldEvents', () => {
    describe('no saved data', () => {
      it('returns 0', async () => {
        mockQuery([]);
        const result = await queryVaultYieldEvents(chain, MOCK_VAULT_DEFINITION);
        expect(result).toEqual([]);
      });
    });
    describe('previously stored harvest data', () => {
      it('returns relevant yield events for the current yield period', async () => {
        const expected = [MOCK_YIELD_EVENT];
        mockQuery(expected);
        const result = await queryVaultYieldEvents(chain, MOCK_VAULT_DEFINITION);
        expect(result).toEqual(expected);
      });
    });
  });

  describe('queryVaultHistoricYieldEvents', () => {
    describe('no saved data', () => {
      it('returns 0', async () => {
        mockQuery([]);
        const result = await queryVaultHistoricYieldEvents(chain, MOCK_VAULT_DEFINITION);
        expect(result).toEqual([]);
      });
    });

    describe('previously stored harvest data', () => {
      it('returns all yield events for a vault', async () => {
        const expected = [MOCK_YIELD_EVENT];
        mockQuery(expected);
        const result = await queryVaultHistoricYieldEvents(chain, MOCK_VAULT_DEFINITION);
        expect(result).toEqual(expected);
      });
    });
  });
});
