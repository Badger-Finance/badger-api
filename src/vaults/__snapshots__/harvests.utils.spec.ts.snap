// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`harvests.utils loadYieldEvents encounters an error requesting an standard vault loads yield events from the subgraph 1`] = `
Array [
  Object {
    "amount": 18087.137228369163,
    "apr": 43.08283925122179,
    "balance": 10000000,
    "block": 15396257,
    "earned": 1171558139.6931558,
    "grossApr": 47.869821390246436,
    "timestamp": 1661253900000,
    "token": "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305",
    "type": "Tree Distribution",
    "value": 66170000000,
  },
  Object {
    "amount": 8273.444229204368,
    "apr": 4.084825017457429,
    "balance": 10000000,
    "block": 15396257,
    "earned": 111079262.22129785,
    "grossApr": 4.538694463841588,
    "timestamp": 1661253900000,
    "token": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    "type": "Tree Distribution",
    "value": 66170000000,
  },
  Object {
    "amount": 4261.975371596002,
    "apr": 1.7384462495527824,
    "balance": 10000000,
    "block": 15388629,
    "earned": 47273830.821742855,
    "grossApr": 1.931606943947536,
    "timestamp": 1661148573000,
    "token": "0x2B5455aac8d64C14786c3a29858E43b5945819C0",
    "type": "Tree Distribution",
    "value": 66170000000,
  },
  Object {
    "amount": 4223.102994638644,
    "apr": 1.7225903301631276,
    "balance": 10000000,
    "block": 15363244,
    "earned": 46842658.416531846,
    "grossApr": 1.9139892557368086,
    "timestamp": 1660800880000,
    "token": "0x2B5455aac8d64C14786c3a29858E43b5945819C0",
    "type": "Tree Distribution",
    "value": 66170000000,
  },
  Object {
    "amount": 4285.530143667973,
    "apr": 1.7480541664451443,
    "balance": 10000000,
    "block": 15337891,
    "earned": 47535100.35356516,
    "grossApr": 1.9422824071612717,
    "timestamp": 1660455256000,
    "token": "0x2B5455aac8d64C14786c3a29858E43b5945819C0",
    "type": "Tree Distribution",
    "value": 66170000000,
  },
  Object {
    "amount": 18705.595517466125,
    "apr": 44.55598222107543,
    "balance": 10000000,
    "block": 15314213,
    "earned": 1211617538.4528334,
    "grossApr": 49.50664691230604,
    "timestamp": 1660132736000,
    "token": "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305",
    "type": "Tree Distribution",
    "value": 66170000000,
  },
  Object {
    "amount": 8418.142660217834,
    "apr": 4.156266578506966,
    "balance": 10000000,
    "block": 15314213,
    "earned": 113021983.35608464,
    "grossApr": 4.618073976118851,
    "timestamp": 1660132736000,
    "token": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    "type": "Tree Distribution",
    "value": 66170000000,
  },
  Object {
    "amount": 5603.2033092929305,
    "apr": 2.285528875516314,
    "balance": 10000000,
    "block": 15312417,
    "earned": 62150731.10667718,
    "grossApr": 2.53947652835146,
    "timestamp": 1660107820000,
    "token": "0x2B5455aac8d64C14786c3a29858E43b5945819C0",
    "type": "Tree Distribution",
    "value": 66170000000,
  },
]
`;

exports[`harvests.utils loadYieldEvents requesting an influence vault loads yield events from the subgraph 1`] = `
Array [
  Object {
    "amount": 18087.137228369163,
    "apr": 9.780451982747769,
    "balance": 4500000,
    "block": 15396257,
    "earned": 1171558139.6931558,
    "grossApr": 10.867168869719743,
    "timestamp": 1661253900000,
    "token": "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305",
    "type": "Tree Distribution",
    "value": 291478500000,
  },
  Object {
    "amount": 8273.444229204368,
    "apr": 0.9273166679709072,
    "balance": 4500000,
    "block": 15396257,
    "earned": 111079262.22129785,
    "grossApr": 1.030351853301008,
    "timestamp": 1661253900000,
    "token": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    "type": "Tree Distribution",
    "value": 291478500000,
  },
  Object {
    "amount": 4261.975371596002,
    "apr": 0.39465342497957007,
    "balance": 4500000,
    "block": 15388629,
    "earned": 47273830.821742855,
    "grossApr": 0.4385038055328557,
    "timestamp": 1661148573000,
    "token": "0x2B5455aac8d64C14786c3a29858E43b5945819C0",
    "type": "Tree Distribution",
    "value": 291478500000,
  },
  Object {
    "amount": 4223.102994638644,
    "apr": 0.3910538929865982,
    "balance": 4500000,
    "block": 15363244,
    "earned": 46842658.416531846,
    "grossApr": 0.4345043255406647,
    "timestamp": 1660800880000,
    "token": "0x2B5455aac8d64C14786c3a29858E43b5945819C0",
    "type": "Tree Distribution",
    "value": 291478500000,
  },
  Object {
    "amount": 4285.530143667973,
    "apr": 0.39683456650722165,
    "balance": 4500000,
    "block": 15337891,
    "earned": 47535100.35356516,
    "grossApr": 0.44092729611913517,
    "timestamp": 1660455256000,
    "token": "0x2B5455aac8d64C14786c3a29858E43b5945819C0",
    "type": "Tree Distribution",
    "value": 291478500000,
  },
  Object {
    "amount": 18705.595517466125,
    "apr": 10.114877576111313,
    "balance": 4500000,
    "block": 15314213,
    "earned": 1211617538.4528334,
    "grossApr": 11.238752862345905,
    "timestamp": 1660132736000,
    "token": "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305",
    "type": "Tree Distribution",
    "value": 291478500000,
  },
  Object {
    "amount": 8418.142660217834,
    "apr": 0.9435349759924179,
    "balance": 4500000,
    "block": 15314213,
    "earned": 113021983.35608464,
    "grossApr": 1.048372195547131,
    "timestamp": 1660132736000,
    "token": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    "type": "Tree Distribution",
    "value": 291478500000,
  },
  Object {
    "amount": 5603.2033092929305,
    "apr": 0.5188494029333707,
    "balance": 4500000,
    "block": 15312417,
    "earned": 62150731.10667718,
    "grossApr": 0.5764993365926341,
    "timestamp": 1660107820000,
    "token": "0x2B5455aac8d64C14786c3a29858E43b5945819C0",
    "type": "Tree Distribution",
    "value": 291478500000,
  },
]
`;

exports[`harvests.utils loadYieldEvents requesting an standard vault loads yield events from the subgraph 1`] = `
Array [
  Object {
    "amount": 6462.338060585159,
    "apr": 6.006022314128918,
    "balance": 10000000,
    "block": 14719741,
    "earned": 163322669.80516872,
    "grossApr": 6.67335812680991,
    "timestamp": 1651788304000,
    "token": "0x62B9c7356A2Dc64a1969e19C23e4f579F9810Aa7",
    "type": "Harvest",
    "value": 66170000000,
  },
  Object {
    "amount": 414.4540216181148,
    "apr": 0.9872129439251023,
    "balance": 10000000,
    "block": 14719741,
    "earned": 26845430.342270147,
    "grossApr": 1.0969032710278914,
    "timestamp": 1651788304000,
    "token": "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305",
    "type": "Tree Distribution",
    "value": 66170000000,
  },
  Object {
    "amount": 10313.848518739553,
    "apr": 9.585571625524985,
    "balance": 10000000,
    "block": 14694100,
    "earned": 260661893.61410472,
    "grossApr": 10.650635139472206,
    "timestamp": 1651437281000,
    "token": "0x62B9c7356A2Dc64a1969e19C23e4f579F9810Aa7",
    "type": "Harvest",
    "value": 66170000000,
  },
  Object {
    "amount": 421.1186708551738,
    "apr": 1.003087872506704,
    "balance": 10000000,
    "block": 14694100,
    "earned": 27277119.667302173,
    "grossApr": 1.1145420805630046,
    "timestamp": 1651437281000,
    "token": "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305",
    "type": "Tree Distribution",
    "value": 66170000000,
  },
  Object {
    "amount": 8548.95779356142,
    "apr": 7.94530258078559,
    "balance": 10000000,
    "block": 14668590,
    "earned": 216057810.31667775,
    "grossApr": 8.828113978650656,
    "timestamp": 1651090393000,
    "token": "0x62B9c7356A2Dc64a1969e19C23e4f579F9810Aa7",
    "type": "Harvest",
    "value": 66170000000,
  },
  Object {
    "amount": 423.486896366854,
    "apr": 1.0087288911898782,
    "balance": 10000000,
    "block": 14668590,
    "earned": 27430516.738370236,
    "grossApr": 1.1208098790998646,
    "timestamp": 1651090393000,
    "token": "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305",
    "type": "Tree Distribution",
    "value": 66170000000,
  },
  Object {
    "amount": 6756.015534480465,
    "apr": 6.278962764603047,
    "balance": 10000000,
    "block": 14642977,
    "earned": 170744780.6029248,
    "grossApr": 6.9766252940033855,
    "timestamp": 1650742506000,
    "token": "0x62B9c7356A2Dc64a1969e19C23e4f579F9810Aa7",
    "type": "Harvest",
    "value": 66170000000,
  },
  Object {
    "amount": 451.8604821337721,
    "apr": 1.076313640459075,
    "balance": 10000000,
    "block": 14642977,
    "earned": 29268359.00925082,
    "grossApr": 1.1959040449545277,
    "timestamp": 1650742506000,
    "token": "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305",
    "type": "Tree Distribution",
    "value": 66170000000,
  },
  Object {
    "amount": 6229.096493405911,
    "apr": 5.7892502969537425,
    "balance": 10000000,
    "block": 14617341,
    "earned": 157427955.6778476,
    "grossApr": 6.432500329948603,
    "timestamp": 1650394973000,
    "token": "0x62B9c7356A2Dc64a1969e19C23e4f579F9810Aa7",
    "type": "Harvest",
    "value": 66170000000,
  },
  Object {
    "amount": 416.1743051520761,
    "apr": 0.991310590668443,
    "balance": 10000000,
    "block": 14617341,
    "earned": 26956858.267615426,
    "grossApr": 1.1014562118538256,
    "timestamp": 1650394973000,
    "token": "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305",
    "type": "Tree Distribution",
    "value": 66170000000,
  },
]
`;
